File: projects/28-telecom-tower-monitoring/infra/verify_pipeline.js
// End-to-end local check: poll DynamoDB until every tower signal has a stored window.
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";

const ENDPOINT = process.env.AWS_ENDPOINT_URL || "http://localhost:4580";
const REGION = process.env.AWS_REGION || "eu-west-1";
const TABLE = process.env.TABLE_NAME || "ctm-readings";
const TIMEOUT = Number(process.env.VERIFY_TIMEOUT || 90) * 1000;

const SIGNALS = ["dc_load_amps", "battery_charge_pct", "genset_fuel_pct", "cabinet_temp_c", "rf_utilization_pct"];

const doc = DynamoDBDocumentClient.from(new DynamoDBClient({
  endpoint: ENDPOINT,
  region: REGION,
  credentials: { accessKeyId: "test", secretAccessKey: "test" },
}));

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function storedSignals() {
  const found = new Set();
  for (const signal of SIGNALS) {
    const res = await doc.send(new QueryCommand({
      TableName: TABLE,
      KeyConditionExpression: "sensor_type = :s",
      ExpressionAttributeValues: { ":s": signal },
      Limit: 1,
    }));
    if (res.Items && res.Items.length > 0) found.add(signal);
  }
  return found;
}

async function main() {
  const deadline = Date.now() + TIMEOUT;
  while (Date.now() < deadline) {
    const found = await storedSignals();
    console.log("stored signals:", [...found].sort().join(", ") || "(none)");
    if (found.size === SIGNALS.length) {
      console.log("PASS: every signal reached DynamoDB");
      return;
    }
    await sleep(3000);
  }
  console.error("FAIL: not all signals stored within the timeout");
  process.exit(1);
}

main();

File: projects/27-river-flood-monitoring/infra/verify_pipeline.py
"""End-to-end local check: poll DynamoDB until every flood signal has at least one stored window."""
import os
import sys
import time

import boto3
from boto3.dynamodb.conditions import Key

ENDPOINT = os.getenv("AWS_ENDPOINT_URL", "http://localhost:4579")
REGION = os.getenv("AWS_REGION", "eu-west-1")
TABLE = os.getenv("TABLE_NAME", "rfw-readings")
TIMEOUT = int(os.getenv("VERIFY_TIMEOUT", "90"))

SIGNALS = ["river_level_m", "rainfall_mmph", "flow_velocity_ms", "soil_moisture_pct", "turbidity_ntu"]

_table = boto3.resource("dynamodb", endpoint_url=ENDPOINT, region_name=REGION,
                        aws_access_key_id="test", aws_secret_access_key="test").Table(TABLE)


def _stored_signals():
    found = set()
    for signal in SIGNALS:
        response = _table.query(KeyConditionExpression=Key("sensor_type").eq(signal), Limit=1)
        if response["Items"]:
            found.add(signal)
    return found


def main():
    deadline = time.monotonic() + TIMEOUT
    while time.monotonic() < deadline:
        found = _stored_signals()
        print(f"stored signals: {sorted(found)}", flush=True)
        if len(found) == len(SIGNALS):
            print("PASS: every signal reached DynamoDB", flush=True)
            return
        time.sleep(3)
    print("FAIL: not all signals stored within the timeout", file=sys.stderr)
    sys.exit(1)


if __name__ == "__main__":
    main()
